mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java

```diff
@@ -695,24 +773,187 @@ private void createUpdateEntityInfoIndexes(EntityInfo ei) {
-            throw e;
         } catch (Exception e) {
             throw analyzeException(e, MetadataConstants.ERR_ILL_FORMED_METADATA);
         } finally {
             Error.pop();
         }
-
         LOGGER.debug("createUpdateEntityInfoIndexes: end");
     }
 
+    /**
+     * Reindexes all lightblue managed, hidden, indexes.
+     *
+     * This operation is blocking and is therefore suggested to be ran in a separate thread
+     *
+     * @param md
+     * @throws IOException
+     */
+    public void reindex(EntityMetadata md) throws IOException {
+        Map<String, String> fieldMap = Translator.getCaseInsensitiveIndexes(md.getEntityInfo().getIndexes().getIndexes())
+                .collect(Collectors.toMap(
+                        i -> i.getField().toString(),
+                        i -> Translator.getHiddenForField(i.getField()).toString()));
+        populateHiddenFields(md.getEntityInfo(), fieldMap);
+    }
+
+    /**
+     *
+     * Populates all hidden fields from their initial index values in the collection in this context
+     *
+     * This method has the potential to be extremely heavy and nonperformant.  Recommended to run in a background thread.
+     *
+     * @param ei
+     * @param fieldMap <index, hiddenPath>
+     * @throws IOException
+     * @throws URISyntaxException
+     */
+    protected void populateHiddenFields(EntityInfo ei, Map<String, String> fieldMap) throws IOException {
+        LOGGER.debug("Starting population of hidden fields due to new or modified indexes.");
+        MongoDataStore ds = (MongoDataStore) ei.getDataStore();
+        DB entityDB = dbResolver.get(ds);
+        DBCollection coll = entityDB.getCollection(ds.getCollectionName());
+        DBCursor cursor = coll.find();
+
+        while (cursor.hasNext()) {
+            DBObject doc = cursor.next();
+            DBObject original = (DBObject) ((BasicDBObject) doc).copy();
+            for (String index : fieldMap.keySet()) {
+                int arrIndex = index.indexOf("*");
+                if (arrIndex > -1) {
+                    // recurse if we have any arrays
+                    doMultiArrayMap(doc, index, fieldMap.get(index), arrIndex);
+                } else {
+                    Object dbObject = Translator.getDBObject(doc, new Path(index));
+                    ObjectNode arrNode = JsonNodeFactory.instance.objectNode();
+                    JsonDoc.modify(arrNode, new Path(fieldMap.get(index)), JsonNodeFactory.instance.textNode(dbObject.toString().toUpperCase()), true);
+                    ObjectMapper mapper = new ObjectMapper();
+                    JsonNode merged = merge(mapper.readTree(doc.toString()), mapper.readTree(arrNode.toString()));
+                    DBObject obj = (DBObject) JSON.parse(merged.toString());
+                    ((BasicDBObject) doc).clear();
+                    ((BasicDBObject) doc).putAll(obj);
+                }
+            }
+            if (!doc.equals(original)) {
```

Comment:
Doc says so, no need to verify

On Wed, Mar 23, 2016 at 2:24 PM, Brandon Vulaj <notifications@github.com>
wrote:

> In
> mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoCRUDController.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/212#discussion_r57230282>
> :
>
> > +                int arrIndex = index.indexOf("*");
> > +                if (arrIndex > -1) {
> > +                    // recurse if we have any arrays
> > +                    doMultiArrayMap(doc, index, fieldMap.get(index), arrIndex);
> > +                } else {
> > +                    Object dbObject = Translator.getDBObject(doc, new Path(index));
> > +                    ObjectNode arrNode = JsonNodeFactory.instance.objectNode();
> > +                    JsonDoc.modify(arrNode, new Path(fieldMap.get(index)), JsonNodeFactory.instance.textNode(dbObject.toString().toUpperCase()), true);
> > +                    ObjectMapper mapper = new ObjectMapper();
> > +                    JsonNode merged = merge(mapper.readTree(doc.toString()), mapper.readTree(arrNode.toString()));
> > +                    DBObject obj = (DBObject) JSON.parse(merged.toString());
> > +                    ((BasicDBObject) doc).clear();
> > +                    ((BasicDBObject) doc).putAll(obj);
> > +                }
> > +            }
> > +            if (!doc.equals(original)) {
>
> @bserdar <https://github.com/bserdar> Based from the doc
> http://api.mongodb.org/java/current/org/bson/BasicBSONObject.html#equals-java.lang.Object-
>
> I would expect it to. I can try and add a test to verify.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/212/files/12f520cb9681286ad5c05d5639d3baeb75f125b5#r57230282>
>